### PR TITLE
test: show that universal libraries can't share the public_name

### PR DESCRIPTION
**Pull request by anmonteiro**
Inspired by a question from @andreypopp.

This is a bit of a sad limitation, because it means `enabled_if` just proliferates when depending on installed libraries.

### CODE REVIEW

File: test/blackbox-tests/test-cases/enabled_if/eif-public-name-collision.t [hunk 1]
```diff
@@ -0,0 +1,64 @@
+Public libraries using the same `public_name`, in different contexts
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/enabled_if/eif-public-name-collision.t:1]: Could you describe the test a bit more here? It should be possible to know what it's about without knowing any context.

**Comment by jchavarri** [comment]
I tried using `(switch)` in the `melange` context defined in the test so that public libraries get installed in separate switches, but the result is the same. 😕 I guess Dune duplication checks come before actually checking if the public libraries are being installed in separate switches?

> because it means `enabled_if` just proliferates when depending on installed libraries.

I fail to see how the premise of this PR would work. How can we avoid the proliferation of `enabled_if` without something like https://github.com/ocaml/dune/issues/10362? As a consumer of a universal library, how can I use it without having two contexts? In reality, a "universal" library consists of two implementations, so I can't imagine a way in which these two implementations are exposed under a single name. For this, Dune already provides a solution: virtual libraries. But they require to hide the implementation details behind abstract types to make it work. Universal libraries allow you to expose the implementations to consumers, at the cost of defining two build contexts.

**Comment by anmonteiro** [comment]
@rgrinberg thanks, I made the description a bit more clear.

**Review by jchavarri — APPROVED** [review]

File: test/blackbox-tests/test-cases/enabled_if/eif-public-name-collision.t [hunk 2]
```diff
@@ -0,0 +1,65 @@
+Show that public library names can't be defined twice, even in different
+contexts
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/enabled_if/eif-public-name-collision.t:2]: It's worth adding a statement on if this is a bug or an intentional limitation.

**Review by rgrinberg — APPROVED** [review]

File: test/blackbox-tests/test-cases/enabled_if/eif-public-name-collision.t (hunk 2 above)
Comment (anmonteiro) [reply test/blackbox-tests/test-cases/enabled_if/eif-public-name-collision.t:2]: to be honest, we're not sure. It depends a bit on the usage, and how they're installed:

- for Melange libraries that clash, it's not a problem if they have the same public name, as we install melange artifacts under `<lib_name>/melange` anyway. 
- for other libraries, if we decided to install them in the same context, they would be overwritten.

We're working on an installation RFC with the aim of making a special case for a "melange context" that can be installed in the same prefix as native libraries with the same name.

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/enabled_if/eif-public-name-collision.t:2]: Okay, well I think this particular issue is easily fixed. We just ignore libraries with `enabled_if` for the purposes of `libname_conflict_map`.

**Pull request merged**